crates/ingress-kafka/src/builder.rs

```diff
@@ -0,0 +1,319 @@
+// Copyright (c) 2023 - 2025 Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::borrow::Borrow;
+
+use anyhow::bail;
+use base64::Engine;
+use bytes::Bytes;
+use opentelemetry::propagation::{Extractor, TextMapPropagator};
+use opentelemetry::trace::{Span, SpanContext, TraceContextExt};
+use opentelemetry_sdk::propagation::TraceContextPropagator;
+use rdkafka::Message;
+use rdkafka::message::BorrowedMessage;
+use tracing::{info_span, trace};
+
+use restate_storage_api::deduplication_table::DedupInformation;
+use restate_types::identifiers::{InvocationId, InvocationUuid, WithPartitionKey, partitioner};
+use restate_types::invocation::{Header, InvocationTarget, ServiceInvocation, SpanRelation};
+use restate_types::live::{self, Live};
+use restate_types::schema::Schema;
+use restate_types::schema::invocation_target::{DeploymentStatus, InvocationTargetResolver};
+use restate_types::schema::subscriptions::{EventInvocationTargetTemplate, Sink, Subscription};
+use restate_wal_protocol::{Command, Destination, Envelope, Source};
+
+use crate::Error;
+
+#[derive(Clone)]
+pub struct EnvelopeBuilder {
+    subscription: Subscription,
+    schema: Live<Schema>,
+    subscription_id: String,
```

Comment:
Maybe add a comment why we are storing the subscription_id explicitly even though we have `subscription` available here.